crates/arroyo-connectors/src/nats/sink/mod.rs

```diff
@@ -0,0 +1,97 @@
+use super::get_nats_client;
+use super::NatsConfig;
+use super::NatsTable;
+use arrow::array::RecordBatch;
+use arroyo_formats::ser::ArrowSerializer;
+use arroyo_operator::context::ArrowContext;
+use arroyo_operator::operator::ArrowOperator;
+use arroyo_rpc::grpc::TableConfig;
+use arroyo_rpc::ControlMessage;
+use arroyo_rpc::ControlResp;
+use arroyo_types::*;
+use async_trait::async_trait;
+use std::collections::HashMap;
+use tracing::warn;
+
+pub struct NatsSinkFunc {
+    pub subject: String,
+    pub servers: String,
+    pub connection: NatsConfig,
+    pub table: NatsTable,
+    pub publisher: Option<async_nats::Client>,
+    pub client_config: HashMap<String, String>,
+    pub serializer: ArrowSerializer,
+}
+
+#[async_trait]
+impl ArrowOperator for NatsSinkFunc {
+    fn name(&self) -> String {
+        format!("nats-publisher-{}", self.subject)
+    }
+
+    fn tables(&self) -> HashMap<String, TableConfig> {
+        HashMap::new()
+    }
+
+    async fn on_start(&mut self, _ctx: &mut ArrowContext) {
+        match get_nats_client(&self.connection, &self.table).await {
+            Ok(client) => {
+                self.publisher = Some(client);
+            }
+            Err(e) => {
+                panic!("Failed to construct NATS publisher: {:?}", e);
+            }
+        }
+    }
+
+    async fn on_close(&mut self, _: &Option<SignalMessage>, ctx: &mut ArrowContext) {
+        if let Some(ControlMessage::Commit { epoch, commit_data }) = ctx.control_rx.recv().await {
+            self.handle_commit(epoch, &commit_data, ctx).await;
+        } else {
+            warn!("No commit message received, not committing")
+        }
+    }
+
+    async fn handle_checkpoint(&mut self, _: CheckpointBarrier, _ctx: &mut ArrowContext) {
+        // TODO: Is it necessary to insert any kind of checklpoint in the state for NATS sink?
```

Comment:
Short version: checkpointing in the sink isn't necessary for correctness; it's fine to just flush all in-progress messages. 

Longer version: there are two main strategies to achieve at-least once in a sink. The first (and usually simpler one) is what you've done here: flush the messages before finishing the checkpoint; this ensures that we won't drop any data because the downstream system will have acknowledged our writes before we mark the checkpoint as complete. If we fail during this process, we recover from the previous checkpoint and will replay the data that potentially got dropped.

The other strategy is to checkpoint our in-progress data. The downside of this is that we potentially have to write a lot more data to state. The upside is that we don't depend on the downstream system to be available in order to checkpoint. So if nats is down for 30 seconds, we're able to continue processing without blocking checkpointing.

So there's no one right answer, and often sinks will offer both options depending on which tradeoff the user prefers.